Restructure NEST 3.0 guides

jessica-mitchell opened the pull request:
Due to the difficulty of navigating the very long NEST 3.0 guide, this PR re-organizes the NEST 3.0 guides by splitting it up into sections.
Basically the one file that contains the comparison between functions of NEST 2.x and NEST 3.0 remains the same
The other file was split such that
* spatially-structured related content moved to the directory guides/spatial/
* a section called NEST3.0 features includes, as separate files, new functionality of nodes, new functionality of handling connections, parametrization and recording from simulations
* The what's removed from NEST 3.0 becomes its own file as well

In addition the wording on the homepage is changed and hopefully more helpful.

There are still 2 pieces of information that have not been re-added, as I was not sure where to put them @stinebuu @hakonsbm  maybe you have some suggestions?:

1. Use allow_offgrid_times throughout: In the model ``spike_generator``, the parameter ``allow_offgrid_spikes`` is renamed                                 
  ``allow_offgrid_times`` for consistency with other models.                                                          
                                                                                                                      
2.  Use unit ms instead of number of simulation steps:The ``structural_plasticity_update_interval`` now has the unit ms instead of                                        
  number of simulation steps.                                                                                         
                              

To see how this looks, you can see it here: https://nest-test.readthedocs.io/en/separate-nest3-guides/index.html
@jougs @terhorstd Let me know if you find this easier to navigate and find information

hakonsbm commented:
@jessica-mitchell I don't think they really fit in any of the guides. Maybe they need their own section? Or perhaps they can be below the guide links in the `All about NEST 3.0` section?

stinebuu commented:
@jessica-mitchell Another option is to create a new table entry in the Reference guide: NEST 2.x vs NEST 3.0 page and place it here. I do not think it will impact that many users, so I think it would be ok to just have it in the comparison guide.

stinebuu commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,457 @@
+    positions = nest.spatial.free([[x, 0.] for x in np.linspace(0, 1.0, N)])
+    spatial_nodes = nest.Create('iaf_psc_alpha', positions=positions)
+
+    parameter = nest.spatial_distributions.exponential(nest.spatial.distance, beta_Ca=0.15)
```
```suggestion
    parameter = nest.spatial_distributions.exponential(nest.spatial.distance, beta=0.15)
```

stinebuu reviewed (changes requested):
Thanks @jessica-mitchell! I ended up reading through all the guides, and found that some of it is a bit outdated, so I added quite alot of comments, but they shouldn't be too hard to address I think.

jessica-mitchell commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/nest3_node_handles.rst:
```diff
@@ -0,0 +1,382 @@
+will point out which parameters can be set and which are read-only.
+
+
+Dictionary with lists when setting parameters
```
Should this section stay on this page then? Or is there a better place for it (like maybe the parametrization.rst file?)

stinebuu replied:
I think this is the best page for it, parametrization.rst is slightly wrong as well. I think it is ok, I like it better now that the connect_arrays section is gone I think.

jessica-mitchell commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/index.rst:
```diff
@@ -0,0 +1,10 @@
+NEST 3.0 Features
+=================
+
```
I added a little line, but I don't know what to say that hasn't been said in other pages (homepage and nest2_to_nest3/index.rst)

jessica-mitchell commented on doc/userdoc/tutorials/pynest_tutorial/part_4_spatially_structured_networks.rst:
```diff
@@ -375,7 +375,7 @@ was built correctly:
 
 It may also be useful to look at the ``.spatial`` property of the
 NodeCollection, which describes the spatial properties. Other useful
-functions that may be of help are listed in the :doc:`../../guides/nest2_to_nest3/nest2_to_nest3_detailed_transition_guide`.
+functions that may be of help are listed in the :doc:`../../guides/nest2_to_nest3/refguide_nest2_nest3`.
```
I agree - last line was removed

jougs commented on doc/userdoc/guides/nest2_to_nest3/whats_removed.rst:
```diff
@@ -0,0 +1,81 @@
+for a full list of functions that have changed):
+
+- BeginSubnet
+- ChangeSubnet
```
```suggestion
- CGConnect
- ChangeSubnet
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/whats_removed.rst:
```diff
@@ -0,0 +1,81 @@
+  | iaf_psc_alpha_presc                          | iaf_psc_alpha_ps                              |
+  +----------------------------------------------+-----------------------------------------------+
+  | iaf_psc_delta_canon                          | iaf_psc_delta_ps                              |
+  +----------------------------------------------+-----------------------------------------------+
```
Please add the name change of `spike_detector` to `spike_recorder`

jessica-mitchell replied:
done

jougs commented on doc/userdoc/guides/nest2_to_nest3/whats_removed.rst:
```diff
@@ -0,0 +1,81 @@
+Functions
+~~~~~~~~~
+
+Some functions have also been removed. The removed functions where either related to subnets,
+or they can be replaced by using other functions with indexing into a NodeCollection.
+The removed functions are (see also :doc:`refguide_nest2_nest3`
+for a full list of functions that have changed):
```
I think this list would massively benefit if the items in the list were linked to related sections in `refguide_nest2_nest3` or the corresponding section of a sub-page under `nest3_features`

jessica-mitchell replied:
Links added

jougs reviewed (changes requested):
Many thanks! I really think these changes improve the readability of the conversion guides a lot!

As I'm a bit short on time right now, I only browsed through the changeset and have added some comments inline whenever I noticed something. I'll try to have a second look by the end of the week. If I don't, please feel free to ping me.

sarakonradi commented on doc/userdoc/guides/nest2_to_nest3/index.rst:
```diff
@@ -0,0 +1,20 @@
+.. toctree::
+   :maxdepth: 1
+
+   Find out what new features NEST 3.0 has to offer <nest3_features/index>
+   Everything you need to know about spatially-structured networks (topology module) <../spatial/index>
+   Compare old versus new function syntax in our reference guide <refguide_nest2_nest3>
+   whats_removed
```
Should we make these more skimmable? For example:

- New features
- Spatially-structured networks (topology module)
- Old versus new function syntax
- What's removed?

sarakonradi commented on doc/userdoc/guides/nest2_to_nest3/index.rst:
```diff
@@ -0,0 +1,20 @@
+.. toctree::
+   :maxdepth: 1
+
+   Find out what new features NEST 3.0 has to offer <nest3_features/index>
+   Everything you need to know about spatially-structured networks (topology module) <../spatial/index>
+   Compare old versus new function syntax in our reference guide <refguide_nest2_nest3>
+   whats_removed
```
Should we make these more skimmable? For example:

- New features
- Spatially-structured networks (topology module)
- Old versus new function syntax
- What's removed?

jessica-mitchell replied:
Good point, done!

sarakonradi reviewed (changes requested):
@jessica-mitchell Thanks a lot for this! Just added a minor comment.

stinebuu commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/nest3_node_handles.rst:
```diff
@@ -0,0 +1,353 @@
+In NEST 3.0, ``nest.Create()`` returns a *NodeCollection* object instead of a list of global IDs.
+This provides a more compact and flexible way for handling nodes.
+
+In most use cases, you will not need to make any changes to your scripts in NEST 3.0, unless you have used **topology** or **subnets**.
```
I think many changes is better now, as people need to change to `synapse_model` and models like `spike_detector` is renamed.
```suggestion
In most use cases, you will not need to make many changes to your scripts in NEST 3.0, unless you have used **topology** or **subnets**.
```

stinebuu reviewed (approved):
Thanks @jessica-mitchell, the guides are a lot more readable now. I just have a tiny comment, but other than that I am happy.

jessica-mitchell commented:
> Thanks @jessica-mitchell, the guides are a lot more readable now. I just have a tiny comment, but other than that I am happy.

@stinebuu I also want to incorporate the random number generator docs once that PR is merged, so I'll need another review for those changes

jessica-mitchell commented:
@jougs @sarakonradi @stinebuu  I added the random number generator doc into the new strucutre. Not much has changed, the section in the old overview file is now its own file in nest3_features and is included in the index.
I updated some headings to follow the style guide.

stinebuu commented on doc/userdoc/index.rst:
```diff
@@ -7,11 +7,12 @@ Welcome to the NEST simulator documentation!
+You can find all the information in our section :doc:`guides/nest2_to_nest3/index`.
 
-- Visit our :doc:`What's new? <guides/nest2_to_nest3/nest2_to_nest3_overview>` guide to get an overview of NEST 3.0 and the new functionality.
+**Note that some of your simulations may need to be updated to run in NEST 3.0!**
```
I think maybe it should be `scripts` instead of `simulations`. Maybe `simulation scripts`.

jessica-mitchell replied:
Done!

jessica-mitchell commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/nest3_node_handles.rst:
```diff
@@ -0,0 +1,352 @@
+
+    >>>  print(nrns[2:9:3])
+         NodeCollection(metadata=None,
+                       model=iaf_psc_alpha, size=2, first=3, last=9, step=3)
```
Yes, it should have another space, I'll fix it.

jessica-mitchell commented:
@jougs Could you have a final look at this PR? Thanks!

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+.. image:: ../../../static/img/NEST3_24_0.png
+  :width: 500px
+
+After you have created your spatially distributed nodes, you can use `spatial` property to set
```
```suggestion
After you have created your spatially distributed nodes, you can use the `spatial` property to set
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+
+  These parameters represent positions of neurons or distances between two
+  neurons. To set node parameters, only the node position can be used. The
+  others can only be used when connecting.
```
```suggestion
  others can be used when connecting.
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+  | ::                                                 |                                                     |
+  |                                                    |                                                     |
+  |     nest.math.min(x, value)                        | If a value from the Parameter is above a threshold, |
+  |                                                    | the value is replaced with the value of the         |
```
```suggestion
  |                                                    | x, the value is replaced with the value of the      |
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+  +----------------------------------------------------+-----------------------------------------------------+
+  | ::                                                 |                                                     |
+  |                                                    |                                                     |
+  |     nest.math.max(x, value)                        | If a value from the parameter is beneath a          |
+  |                                                    | threshold, the value is replaced with the value of  |
+  |                                                    | the threshold.                                      |
```
```suggestion
  |     nest.math.max(x, value)                        | If a value from the parameter is below a            |
  |                                                    | threshold x, the value is replaced with the value      |
  |                                                    | of the threshold.                                   |
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+
+The ``nest.math.min()`` and ``nest.math.max()`` functions are used to clip
+a parameter. Essentially they work like the standard ``min()`` and
+``max()`` functions, ``nest.math.min()`` yielding the smallest of two
+values, and ``nest.math.max()`` yielding the largest of two values.
```
```suggestion
``max()`` functions, ``nest.math.min()`` yielding the smaller of two
values, and ``nest.math.max()`` yielding the larger of two values.
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+::
+
+    # This yields values between 0.0 and 0.5, where values from the
+    # distribution that are above 0.5 gets set to 0.5.
```
```suggestion
    # distribution that are above 0.5 get set to 0.5.
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+    nest.math.min(nest.random.uniform(), 0.5)
+
+    # This yields values between 0.5 and 1.0, where values from the
+    # distribution that are below 0.5 gets set to 0.5.
```
```suggestion
    # distribution that are below 0.5 get set to 0.5.
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+    nest.math.max(nest.random.uniform(), 0.5)
+
+    # This yields values between 0.2 and 0.7, where values from the
+    # distribution that are smaller than 0.2 or larger than 0.7 gets
```
```suggestion
    # distribution that are smaller than 0.2 or larger than 0.7 get
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+statement. Three arguments are required:
+
+- The first argument is a condition.
+- The second argument is the resulting value or parameter evalued if the
```
```suggestion
- The second argument is the resulting value or parameter evaluated if the
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,454 @@
+- The first argument is a condition.
+- The second argument is the resulting value or parameter evalued if the
+  condition evaluates to true.
+- The third argument is the resulting value or parameter evalued if the
```
```suggestion
- The third argument is the resulting value or parameter evaluated if the
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/random_number_generators.rst:
```diff
@@ -0,0 +1,60 @@
+All random number generators managed by the NEST kernel are now seeded by
+providing a single seed :math:`s`  with :math:`1\leq s \leq 2^{31}-1`. The
+kernel automatically seeds the random number streams on the various parallel
+processes:
```
```suggestion
threads and processes:
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/random_number_generators.rst:
```diff
@@ -0,0 +1,60 @@
+function to N.
+
+The Random123 library is included in NEST, and random number generator types like Philox and
+Threefry are automatically made available.
```
```suggestion
Threefry are available out of the box.
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/recording_simulations.rst:
```diff
@@ -0,0 +1,94 @@
+rationale behind this is that the device is actually not detecting the
+occurence of spikes, but rather only records them. Moreover, the new
+name is more consistent with the naming of other similar devices that
+also end in the suffix `_recorder`.
```
```suggestion
also end in `_recorder`.
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/recording_simulations.rst:
```diff
@@ -0,0 +1,94 @@
+
+Previously, the content and formatting of any output created by a
+recording device could be configured in a fine-grained fashion using
+flags like `withgid`, `withtime`, `withweight`, `withport` and so
+on. In many cases, this, however, lead to a confusing variety of
```
```suggestion
flags like `withgid`, `withtime`, `withweight`, `withport` and some
more. In many cases, this, however, lead to a confusing variety of
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/refguide_nest2_nest3.rst:
```diff
@@ -297,6 +310,9 @@ Functions related to simulation
 
+   .. note::
+
+     The structural_plasticity_update_interval now has the unit `ms` instead of `number of simulation steps`.
```
```suggestion
     The structural_plasticity_update_interval is now reported as time in ms instead of as the number of simulation steps.
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/whats_removed.rst:
```diff
@@ -0,0 +1,92 @@
+Subnets
+~~~~~~~
+
+Subnets are gone. Instead NodeCollections should be used to organize neurons.
```
```suggestion
Subnets are gone. Instead NodeCollections should be used to group and organize neurons.
```

jougs commented on doc/userdoc/guides/nest2_to_nest3/whats_removed.rst:
```diff
@@ -0,0 +1,92 @@
++----------------------+------------------------------+
+| - ResetNetwork       |  See :ref:`sim_ref`          |
++----------------------+------------------------------+
+| - DataConnect        |                              |
```
```suggestion
| - DataConnect        |                              |
| - CGConnect          |                              |

```

jougs commented on doc/userdoc/guides/spatial/overview_spatial.rst:
```diff
@@ -0,0 +1,262 @@
+Create spatially distributed nodes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Creating spatially distributed nodes is now handled with the standard ``nest.Create()`` function.
```
```suggestion
Spatially distributed nodes can now be created using the standard ``nest.Create()`` function.
```

jougs commented on doc/userdoc/guides/spatial/overview_spatial.rst:
```diff
@@ -0,0 +1,262 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Creating spatially distributed nodes is now handled with the standard ``nest.Create()`` function.
+Arguments of node creation have also been changed to make creating
+populations with and without spatial information more unified. To create
```
```suggestion
The arguments of this function have been changed to unify the
creation of populations with and without spatial information. To create
```

jougs commented on doc/userdoc/guides/spatial/overview_spatial.rst:
```diff
@@ -0,0 +1,262 @@
+NodeCollections with two exceptions:
+
+- They cannot be merged, as concatenating NodeCollections with metadata is
+  not allowed.
```
```suggestion
  not supported.
```

jougs commented on doc/userdoc/guides/spatial/overview_spatial.rst:
```diff
@@ -0,0 +1,262 @@
+Connect spatially distributed nodes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Similar to creating nodes with spatial distributions, connecting is now done with the
```
```suggestion
Similar to creating nodes with spatial distributions, such nodes are now connected using the
```

jougs commented on doc/userdoc/guides/spatial/overview_spatial.rst:
```diff
@@ -0,0 +1,262 @@
+  ``use_on_source`` here refers to whether the mask and connection probability
+  should be applied to the source neuron or the target neuron.
+  This is only required for ``pairwise_bernoulli``, as ``fixed_indegree``
+  and ``fixed_outdegree`` implicitly states if we are using the source or
+  target nodes.
```
```suggestion
  and ``fixed_outdegree`` implicitly only apply to either the source or target nodes.
```

jougs commented on doc/userdoc/guides/spatial/overview_spatial.rst:
```diff
@@ -0,0 +1,262 @@
+Masks
+^^^^^
+In NEST 3.0, the mask ``volume`` got removed, as the same mask was already available under the name ``box``.
+The former was only an alias available in NEST for backward compatibility.
```
```suggestion
The former was only an alias available for backward compatibility.
```

jougs commented on doc/userdoc/guides/spatial/overview_spatial.rst:
```diff
@@ -0,0 +1,262 @@
+between the source-target pair at *indx*.
+
+Calling ``.distance`` on a SynapseCollection where either the source or target, or both, are not spatially
+distributed also works, you will receive `nan` whenever one of the nodes is non-spatial.
```
```suggestion
distributed also works, and you will receive `nan` whenever one of the nodes is non-spatial.
```

jougs reviewed (changes requested):
Voilà!

jessica-mitchell commented:
@jougs your suggestions have been applied! thanks

jougs commented on doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst:
```diff
@@ -0,0 +1,457 @@
+    ax[0].set_ylabel('V_m');
+
+
+.. image:: ../../../_static/img/NEST3_13_0.png
```
Is this resolved?

jougs reviewed (approved):
Many thanks!
